internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -0,0 +1,469 @@
+		go ticker.start()
+		tickers = append(tickers, ticker)
+	}
+	return tickers
+}
+
+func newBucketTicker(parent *batcher, tickTime time.Duration) *bucketTicker {
+	return &bucketTicker{
+		ticker:       time.NewTicker(tickTime),
+		nodes:        make(map[string]bool),
+		parent:       parent,
+		pendingNodes: make(chan string, tickerPendingNodesBuffer),
+		stopCn:       make(chan struct{}),
+	}
+}
+
+func (bt *bucketTicker) add(bucketID string) {
+	bt.pendingNodes <- bucketID
+}
+
+func (bt *bucketTicker) start() {
+	bt.once.Do(func() {
+		for {
+			select {
+			case <-bt.ticker.C:
+				for nbKey := range bt.nodes {
+					nb := bt.parent.getBucket(nbKey)
+					// Need to check nil here incase the node was deleted from the parent batcher, but
+					// not deleted from bt.nodes
+					if nb == nil {
+						// re-delete just in case
+						delete(bt.nodes, nbKey)
+						continue
+					}
+					// atomic.LoadPointer only takes unsafe.Pointer interfaces. We do not use unsafe
+					// to skirt around the golang type system.
+					ptr := (*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))
+					b := (*batch)(atomic.LoadPointer(ptr))
+					if atomic.LoadInt64(&nb.lastSent)+nb.timeout.Nanoseconds() > time.Now().UnixNano() {
+						// In this case, we sent more recently than the timeout, so do nothing
+						continue
+					} else if b.getCurrentItemCount() > 0 {
+						// If the batch is non-empty, go ahead and cut it
+						statsTags := processor.StatsTagsForBatch(
+							nb.parent.name, processor.ServiceNameForNode(nb.node), nb.spanFormat,
+						)
+						stats.RecordWithTags(context.Background(), statsTags, statTimeoutTriggerSend.M(1))
+						nb.cutBatch(b)
+					} else {
+						cyclesUntouched := atomic.AddUint32(&nb.cyclesUntouched, 1)
+						if cyclesUntouched > nb.parent.removeAfterCycles {
+							nb.parent.removeBucket(nbKey)
+							delete(bt.nodes, nbKey)
+							// At this point, add() should no longer be callable from the parent,
+							// so mark this nodeBatcher as dead (for adds that have already been called and are racing)
+							// and try to cut a final batch (just in case something was added in between
+							// the above if and marking this nodeBatcher as dead)
+							atomic.StoreUint32(&nb.dead, nodeDead)
+							// Reload batch in case `add` loaded a new batch that was then added to.
+							currBatchPtr := (*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))
+							b = (*batch)(atomic.LoadPointer(currBatchPtr))
+							if b.getCurrentItemCount() > 0 {
+								nb.cutBatch(b)
+							}
+						}
+					}
+				}
+			case newBucketKey := <-bt.pendingNodes:
+				bt.nodes[newBucketKey] = true
+			case <-bt.stopCn:
+				bt.ticker.Stop()
+				return
+			}
+		}
+	})
+}
+
+func (bt *bucketTicker) stop() {
+	close(bt.stopCn)
+}
+
+type batch struct {
+	items         atomic.Value // []*tracepb.Span
+	currCap       uint32
+	nextEmptyItem uint32
+	sendItemsSize uint32
+
+	closed  uint32
+	growMu  sync.Mutex
+	pending int32
```

Comment:
Using atomics might fail due to the arrangement of these fields due to potential alignment issue

Field|64 bit size(bytes)|32 bit size(bytes)
---|---|---
items|16|8
currCap|4|4
nextEmptyItem|4|4
sendItemsSize|4|4
close|4|4
growMu|8|8
pending|4|4

Please examine and adjust accordingly, you might want to ensure that all the mutex is first, then integers and then items.

Reply:
or rather, integers first, then items and then lastly the mutex but just make sure that anything that gets loaded or stored is either on a 4 or 8 alignment boundary.

Reply:
Discussed offline -- using 32 bit ints is safe for both 32 and 64 bit systems since golang keeps stuff word-aligned. Moved `lastSent` to top for alignment since it is 64 bits and therefore needs to be 64-bit aligned for 32 bit systems.